### PR TITLE
Integrated Runner into WorkerServer.

### PR DESCRIPTION
**Pull request by jasonbs10**

### CODE REVIEW

**Comment by jasonbs10** [comment]
TODO: additional testing + unit tests

File: workerapi/client/commands.go [hunk 1]
```diff
@@ -113,3 +113,5 @@ func (c *Client) queryworker(cmd *cobra.Command, args []string) error {
 	log.Println(render.Render(status))
 	return err
 }
+
+//TODO: implement clear()
```
Comment (dbentley) [inline workerapi/client/commands.go]: What is clear?

File: workerapi/server/server.go [hunk 2]
```diff
@@ -3,16 +3,20 @@ package server
...
 	"log"
-	"math/rand"
 	"strconv"
 	"time"
 
 	"github.com/apache/thrift/lib/go/thrift"
 	"github.com/luci/go-render/render"
 	"github.com/scootdev/scoot/common/stats"
+	"github.com/scootdev/scoot/runner"
+	execer "github.com/scootdev/scoot/runner/execer/os"
```
Comment (dbentley) [inline workerapi/server/server.go]: osexecer?

**Comment by jasonbs10** [comment]
I think this is an improvement (although it touched more files than I wanted). Dan, is this ok for you?

File: daemon/client/conn/conn.go [hunk 3]
```diff
@@ -101,25 +102,37 @@ func (c *conn) Echo(arg string) (string, error) {
...
 	r, err := c.client.Status(context.Background(), &protocol.StatusQuery{RunId: string(run)})
 	if err != nil {
-		return runner.ProcessStatus{}, err
+		return runner.ProcessStatus{State: runner.FAILED, Error: err.Error()}
 	}
-	return protocol.ToRunnerStatus(r), nil
+	return protocol.ToRunnerStatus(r)
+}
+
+func (c *conn) StatusAll() []runner.ProcessStatus {
```
Comment (dbentley) [inline daemon/client/conn/conn.go]: BTW, let me just highlight this as a place where I have no idea what the right thing to do is.

Conn wants to offer "hey, you can run this". Options for how to do that:
a) just offer grpc interface
b) create a new domain model
c) be a runner.Runner

All of these seem bad, but for different reasons. So, yeah, clumsy, and I'm sad. What do you think?

(Not a comment on this review, really, but just a conversation to keep in the back of our minds)

File: runner/execer/os/os_execer.go [hunk 4]
```diff
@@ -14,9 +16,13 @@ func NewExecer() execer.Execer {
 type osExecer struct{}
 
 func (e *osExecer) Exec(command execer.Command) (result execer.Process, err error) {
+	if len(command.Argv) <= 0 {
```
Comment (dbentley) [inline runner/execer/os/os_execer.go]: I'd just say == 0; if len starts returning negative numbers, everything is awful.

(But that's so far below a nit...)

File: runner/fake/fake.go [hunk 5]
```diff
@@ -11,12 +11,23 @@ func NewRunner() runner.Runner {
 type fakeRunner struct {
```
Comment (dbentley) [inline runner/fake/fake.go]: Huh, maybe we don't need fake runner at all now that we have the simple runner and the sim execer?

File: runner/local/simple.go [hunk 6]
```diff
@@ -2,69 +2,136 @@ package local
...
-	return result, nil
+	result.Abort()
+	delete(r.running, run)
+	r.runs[run] = runner.AbortStatus(run)
+	return r.runs[run]
 }
 
-func (r *simpleRunner) markFinished(p execer.Process, runId runner.RunId, status execer.ProcessStatus) {
+func (r *simpleRunner) Erase(run runner.RunId) {
+	// Best effort is fine here.
```
Comment (dbentley) [inline runner/local/simple.go]: need to lock

File: runner/local/simple.go [hunk 7]
```diff
@@ -2,69 +2,136 @@ package local
...
+	}
+
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	r.running = nil
+
+	// Check if we already marked this run as done (i.e. in Abort())
+	// Further, check if we aborted due to a timeout.
+	delete(r.running, runId)
+	if r.runs[runId].State.IsDone() {
```
Comment (dbentley) [inline runner/local/simple.go]: It it's already done, shouldn't we not change its status? Like, if it were aborted, it shouldn't later change to timed out?

File: runner/local/simple_test.go [hunk 8]
```diff
@@ -19,27 +21,24 @@ func TestRun(t *testing.T) {
 	assertWait(t, r, firstId, complete(0), "complete 0")
 }
 
-func TestSimul(t *testing.T) {
+func XTestSimul(t *testing.T) {
```
Comment (dbentley) [inline runner/local/simple_test.go]: X? So this isn't running?

File: runner/local/simple.go [hunk 9]
```diff
@@ -2,69 +2,136 @@ package local
...
 	"github.com/scootdev/scoot/runner"
 	"github.com/scootdev/scoot/runner/execer"
-	"sync"
 )
 
 func NewSimpleRunner(exec execer.Execer) runner.Runner {
+	return NewMultiRunner(exec, 1)
+}
+
+func NewMultiRunner(exec execer.Execer, maxRunning int) runner.Runner {
```
Comment (dbentley) [inline runner/local/simple.go]: Why the change to support running multiple simultaneously?

This seems like a big change that makes things more complicated (e.g., we can no longer just have one checkout we reuse, we have to allocate). Is there something it enables right now?

I didn't see tests for this more complex functionality, but I think they're important. Did I miss them, or have you not done them yet?

File: runner/runner.go [hunk 10]
```diff
@@ -85,8 +93,17 @@ type Runner interface {
...
 	// wait a very short period of time for cmd to finish
 	// Run may not wait indefinitely for cmd to finish. This is an async API.
-	Run(cmd *Command) (ProcessStatus, error)
+	Run(cmd *Command) ProcessStatus
 
 	// Status checks the status of run.
-	Status(run RunId) (ProcessStatus, error)
+	Status(run RunId) ProcessStatus
+
+	// Current status of all runs, running and finished.
```
Comment (dbentley) [inline runner/runner.go]: FWIW, I think it's fine if a Runner does throw out some statuses, so maybe say it's best effort for finished runs? WDYT?

File: workerapi/api.go [hunk 11]
```diff
@@ -1,4 +1,89 @@
...
+	if thrift.Error != nil {
+		domain.Error = *thrift.Error
+	}
+	if thrift.ExitCode != nil {
+		domain.ExitCode = int(*thrift.ExitCode)
+	}
+	return domain
+}
+
+func DomainRunStatusToThrift(domain *runner.ProcessStatus) *worker.RunStatus {
```
Comment (dbentley) [inline workerapi/api.go]: Tests?

(When I did this, Caitie suggested property-based testing, which is kinda nifty, but maybe overkill.)

I think table-based testing is easy to do with this kind of input-output so it's not super painful...

File: workerapi/server/server.go [hunk 12]
```diff
@@ -21,99 +24,99 @@ func Serve(handler worker.Worker, addr string, transportFactory thrift.TTranspor
...
-	rs := worker.NewRunStatus()
-	rs.Status = status
-	rs.Info = &info
-	rs.ExitCode = &exitCode
-	return rs
+// Implement thrift worker.Worker interface.
+//
+func (h *handler) QueryWorker() (*worker.WorkerStatus, error) {
+	h.stat.Counter("workerQueries").Inc(1)
+	h.timeLastContact = time.Now()
```
Comment (dbentley) [inline workerapi/server/server.go]: Why do this, instead of just updating the gauge directly right here?

File: workerapi/server/server.go [hunk 13]
```diff
@@ -21,99 +24,99 @@ func Serve(handler worker.Worker, addr string, transportFactory thrift.TTranspor
...
-	wstatus.Running = []string{}
-	for runId, status := range h.runs {
-		if status.Status == worker.Status_RUNNING {
-			wstatus.Running = append(wstatus.Running, runId)
+// Periodically output stats
+func (h *handler) stats() {
+	ticker := time.NewTicker(time.Millisecond * time.Duration(500))
+	for {
+		select {
+		case <-ticker.C:
```
Comment (dbentley) [inline workerapi/server/server.go]: Hrm, this code looks a little meh... It would be nice if the runner could just update the stats as it updates its code.... Probably not worth rejiggering now... But maybe the Runner could also accept a listener that just calls the stats? (so the runner doesn't have to know about stats for, like, a local runner?)

File: workerapi/worker.thrift [hunk 14]
```diff
@@ -1,37 +1,42 @@
...
+  1: required list<RunStatus> runs  # All runs excepting what's been Erase()'d
+  2: optional string versionId      # Version associated with this worker.
+}
+
 struct RunCommand {
-  1: required list<string> argv
-  2: optional string snapshotId
-  3: optional i32 timeoutMs
+  1: required list<string> argv       # Binary followed by any number of arguments.
+  2: optional map<string,string> env  # Binary followed by any number of arguments.
```
Comment (dbentley) [inline workerapi/worker.thrift]: This doc looks wrong....

File: binaries/workerserver/main.go [hunk 15]
```diff
@@ -26,7 +26,7 @@ func main() {
 	protocolFactory := thrift.NewTBinaryProtocolFactoryDefault()
 	transportFactory := thrift.NewTTransportFactory()
 
-	handler := server.NewHandler(stat.Scope("workerserver"))
+	handler := server.DefaultHandler(stat.Scope("workerserver"))
```
Comment (dbentley) [inline binaries/workerserver/main.go]: Any chance I could convince you to undo DefaultHandler and require injecting the runner dependency? I know it looks a little worse right here, but I think it's really good if you have to pass the Runner, because it means that when you're looking at the workerserver, you know it's running something.

(It also means that we don't need to agonize over what the default should be)

**Comment by dbentley** [comment]
LGTM

Yeah, that looks great.

I think the only really blocking comment is about the multiple runner. (If it's important for now, it should be tested)

File: runner/execer/execer.go [hunk 16]
```diff
@@ -30,8 +30,7 @@ type Execer interface {
 type Process interface {
```
Comment (CaitieM20) [inline runner/execer/execer.go]: What is the expected behavior of Wait & Abort.  Comment on what this interface represents and the expected behavior of methods is would be nice.

I assume Wait() is waiting for something to execute.  Should it potentially return an error then? What if Abort is called while we are Waiting?

File: runner/execer/os/os_execer.go [hunk 17]
```diff
@@ -14,9 +16,13 @@ func NewExecer() execer.Execer {
...
+		return nil, errors.New("No command specified.")
```
Comment (CaitieM20) [inline runner/execer/os/os_execer.go]: Should we return a specific error type so the caller can differentiate between bad request and something the execer couldn't do?

File: runner/runner.go [hunk 18]
```diff
@@ -19,14 +19,18 @@ const (
 	// States below are end states
 	// a Process in an end state will not change its state
 
-	// Ran to completion
+	// Run completed, possibly with nonzero exit code.
 	COMPLETE
-	// Could not run to completion
+	// Could not complete, bad request or golang runtime err.
```
Comment (CaitieM20) [inline runner/runner.go]: I think we should differentiate between bad request and failed to run.  Having this distinction will help us abort sending poison pill messages around, or falsely marking nodes as dead

BadRequest means it should not be retried by the caller. 
FailedToRun is a retryable error.

File: workerapi/api.go (hunk 11 above)
Comment (CaitieM20) [inline workerapi/api.go]: Yea whatever is easiest here for testing is nice since the statespace is fairly limited a table test can work just as well!

File: workerapi/server/server.go (hunk 12 above)
Comment (CaitieM20) [inline workerapi/server/server.go]: Agree the typical pattern in most Twitter code is to pass the stats object in, and then just update next to the code its effecting.

This is also really nice when you are debugging code and trying to figure out what statistic names correspond to what part of the code.

File: workerapi/server/server.go (hunk 13 above)
Comment (CaitieM20) [inline workerapi/server/server.go]: Agree why not update the stats next to the code they are tracking thats a more common pattern.

File: workerapi/worker.thrift [hunk 19]
```diff
@@ -1,37 +1,42 @@
...
-//TODO: add useful load information when it comes time to have multiple runs.
-struct WorkerStatus {
-  1: required list<string> running
+  TIMEDOUT = 6
 }
 
 struct RunStatus {
   1: required Status status
-  2: optional string runId
+  2: required string runId
```
Comment (CaitieM20) [inline workerapi/worker.thrift]: Does RunId correspond to task or jobId?  Do these need to be unique?

File: workerapi/worker.thrift [hunk 20]
```diff
@@ -1,37 +1,42 @@
...
   6: optional i32 exitCode
 }
 
+// VersionId will be empty by default. To change this, pass a getVersion() func to server.Handler.
+// - Semantics are up to the implementer, this needn't be a library version, it could be a gitSha, etc.
+//
+// TODO: add useful load information when it comes time to have multiple runs.
+struct WorkerStatus {
+  1: required list<RunStatus> runs  # All runs excepting what's been Erase()'d
+  2: optional string versionId      # Version associated with this worker.
```
Comment (CaitieM20) [inline workerapi/worker.thrift]: what is versionId tracking.  Version of what?

File: workerapi/worker.thrift [hunk 21]
```diff
@@ -1,37 +1,42 @@
...
 }
 
 //TODO: add a method to kill the worker if we can articulate unrecoverable issues.
 service Worker {
-  WorkerStatus QueryWorker()
-  RunStatus Run(1: RunCommand cmd)
-  RunStatus Query(1: string runId)
-  RunStatus Abort(1: string runId)
+  WorkerStatus QueryWorker()         # Overall worker node status.
+  RunStatus Run(1: RunCommand cmd)   # Run a command and return job Status.(RUNNING|FAILED).
```
Comment (CaitieM20) [inline workerapi/worker.thrift]: nit: delete Running|Failed since there are more potential statuses that could be returned

File: daemon/client/conn/conn.go (hunk 3 above)
Comment (CaitieM20) [inline daemon/client/conn/conn.go]: I agree I'm not sure the best way to proceed here, but Im a fan of starting with the simplest thing that works, so the current idea of this just being another implementation of Runner works for now.  And then we can iterate if this becomes problematic.  

It also makes it really clear in code that this is the equivalent to the Runner in the CloudWorker.  Which is a nice parallel to draw.

File: runner/execer/os/os_execer.go (hunk 4 above)
Comment (jasonbs10) [inline runner/execer/os/os_execer.go]: Done.

File: runner/fake/fake.go (hunk 5 above)
Comment (jasonbs10) [inline runner/fake/fake.go]: Let's do that in a new PR, it affects other tests.

File: runner/local/simple.go (hunk 6 above)
Comment (jasonbs10) [inline runner/local/simple.go]: Done.

File: runner/local/simple.go (hunk 7 above)
Comment (jasonbs10) [inline runner/local/simple.go]: Fixed.

File: runner/local/simple_test.go (hunk 8 above)
Comment (jasonbs10) [inline runner/local/simple_test.go]: I forget to re-enable it....

File: runner/runner.go (hunk 10 above)
Comment (jasonbs10) [inline runner/runner.go]: Ehhh, -1, I prefer completeness.

File: workerapi/server/server.go (hunk 12 above)
Comment (jasonbs10) [inline workerapi/server/server.go]: I don't do this at the call sites since I'm measuring an incrementing time and if there's no rpc (which is what I'm measuring), the stats become stale.

File: workerapi/server/server.go (hunk 13 above)
Comment (jasonbs10) [inline workerapi/server/server.go]: Added a TODO to move this into an extended runner.

File: workerapi/worker.thrift (hunk 14 above)
Comment (jasonbs10) [inline workerapi/worker.thrift]: Fixed.

File: binaries/workerserver/main.go (hunk 15 above)
Comment (jasonbs10) [inline binaries/workerserver/main.go]: Absolutely, done.

File: runner/execer/os/os_execer.go (hunk 17 above)
Comment (jasonbs10) [inline runner/execer/os/os_execer.go]: Sure, I've added BADREQUEST, but without the transient connotation.

File: runner/runner.go (hunk 18 above)
Comment (jasonbs10) [inline runner/runner.go]: Updated this comment.

File: workerapi/worker.thrift (hunk 19 above)
Comment (jasonbs10) [inline workerapi/worker.thrift]: Commented. This runId is only unique within each worker

File: workerapi/worker.thrift (hunk 20 above)
Comment (jasonbs10) [inline workerapi/worker.thrift]: Oh, there's a comment above noting this could be sha1 etc.

File: workerapi/worker.thrift (hunk 21 above)
Comment (jasonbs10) [inline workerapi/worker.thrift]: Done.

File: runner/execer/execer.go (hunk 16 above)
Comment (jasonbs10) [inline runner/execer/execer.go]: Added some very basic comments. To elaborate slightly, every combination of calls returns process.FAILED except a solitary call to Wait() when the process exits cleanly.

File: runner/local/simple.go (hunk 9 above)
Comment (dbentley) [inline runner/local/simple.go]: Seems like it's still here?

Comment (jasonbs10) [inline runner/local/simple.go]: I usually push after finishing with comments... I still have to address the missing tests. 






On Sat, Jul 30, 2016 at 11:56 AM -0700, "Dan Bentley" <notifications@github.com> wrote:












In runner/local/simple.go:
>  )
>  
>  func NewSimpleRunner(exec execer.Execer) runner.Runner {
> +	return NewMultiRunner(exec, 1)
> +}
> +
> +func NewMultiRunner(exec execer.Execer, maxRunning int) runner.Runner {



Seems like it's still here?

File: workerapi/api.go (hunk 11 above)
Comment (jasonbs10) [inline workerapi/api.go]: Done.

**Comment by jasonbs10** [comment]
Ok, I think I addressed most of the comments. Let's do a final lgtm if it looks good.

**Comment by dbentley** [comment]
LGTM
